ci/docker/rchain-docker-build-push.sh

```diff
@@ -0,0 +1,142 @@
+else
+    set -e
+fi
+
+if [[ "$#" != "3" ]]; then
+    echo "Invalid amount of parameters."
+    echo "Usage: $0 <branch name> <repo url> <docker hub repo>"
+    echo "Usage: $0 mybranch https://github.com/myrepo/rchain myrepo/rchain:latest"
+    exit
+fi
+
+# Set working branch
+branch_name=$1
+git_repo=$2
+docker_dst_repo=$3
+
+echo "Branch Name: $branch_name"
+echo "Git Repo: $git_repo"
+echo "Docker Repo: $docker_dst_repo"
+echo "Travis Branch: ${TRAVIS_BRANCH}"
+echo "Travis Docker Username: ${DOCKER_USERNAME}"
+echo "5 seconds to cancel if this information is not correct."
+sleep 5 
+
+# Install Docker-CE
+apt update
+apt-get install -y apt-transport-https ca-certificates curl software-properties-common
+curl -fsSL https://download.docker.com/linux/ubuntu/gpg |  apt-key add -
+add-apt-repository    "deb [arch=amd64] https://download.docker.com/linux/ubuntu    xenial    stable"
+apt update
+apt install -y docker-ce
+
+# Get RChain Repo
+apt -y install git
+git clone ${git_repo} 
+cd rchain
+PROJECT_ROOT_DIR=$(pwd -P)
+git checkout ${branch_name} 
+
+### Install all dependencies on Ubuntu 16.04 LTS (Xenial Xerus) for RChain dev environment.
+
+## Verify operating system (OS) version is Ubuntu 16.04 LTS (Xenial Xerus)
+# Add more OS versions as necessary. 
+version=$(cat /etc/*release | grep "^VERSION_ID" | awk -F= '{print $2}' | sed 's/"//g')
+if [[ "$version" == "16.04" ]]; then
+    echo "Running install on Ubuntu 16.04" 
+else
+    echo "Error: Not running on Ubuntu 16.04"
+    echo "Exiting"
+    exit
+fi
+
+## Resynchronize the package index files from their sources
+apt-get update -yqq
+## Install g++ multilib for cross-compiling as rosette is currently only 32-bit
+apt-get install g++-multilib -yqq
+## Install misc tools 
+apt-get install cmake curl git -yqq
+## Install Java OpenJDK 8
+apt-get update -yqq
+#  apt-get install default-jdk -yqq # alternate jdk install 
+apt-get install openjdk-8-jdk -yqq
+
+## Build Needed Crypto
+# Build secp 
+apt-get install autoconf libtool -yqq
+cd ${PROJECT_ROOT_DIR}
+cd crypto
+if [[ -d "secp256k1" ]]; then
+    rm -rf secp256k1 
+fi
+git clone https://github.com/bitcoin-core/secp256k1
+cd secp256k1
+./autogen.sh
+./configure --enable-jni --enable-experimental --enable-module-schnorr --enable-module-ecdh --prefix=/tmp/f/rchain/.tmp
+make
+
+# Build libsodium
+cd ${PROJECT_ROOT_DIR}
+cd crypto
+if [[ -d "libsodium" ]]; then
+    rm -rf libsodium 
+fi
+git clone https://github.com/jedisct1/libsodium --branch stable
+cd libsodium
+./configure
+make && make check
+make install
+
+## Install Haskell Platform
+# ref: https://www.haskell.org/platform/#linux-ubuntu
+# ref: https://www.haskell.org/platform/ # all platforms
+cd ${PROJECT_ROOT_DIR}
+apt-get install haskell-platform -yqq
+
+## Install BNFC Converter 
+# ref: http://bnfc.digitalgrammars.com/
+bnfc_tmp_dir="$(mktemp -dt bnfcbuild.XXXXXX)"
+cd ${bnfc_tmp_dir}
+git clone https://github.com/BNFC/bnfc.git
+cd bnfc/source
+cabal install --global
```

Comment:
It feels like you just included all the code from things like `rchain/scripts/install.sh` and others here. This is going to tend to cause errors when the docker build and other scripts don't match. E.g., if someone updates one but not the other.

Which is authoritative? Is there a reason to maintain both?

Reply:
I should I just split it out. Initially this was just going to be an entirely separate build with "all" code in the script. I was hesitant because of maintaining them in sync but felt it was small and would evolve very soon. I've created this PR https://github.com/rchain/rchain/pull/555 for your review and after that, I will reference the ./scripts/install_secp/sodium/bnf for so I'm only referencing one file on builds. Thanks for the input. My bad on deciding to duplicate and lump. I shouldn't have done that.

Reply:
I did add comment in code clarifying what this does exactly.